lib/enmeshed/connector_api.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+module Enmeshed
+  class ConnectorError < StandardError; end
+
+  class ConnectorApi
+    # The app does not let users scan expired templates but previously scanned and then expired templates can still be submitted,
+    # resulting in the app silently doing nothing. CodeHarbor would still accept Relationships for expired templates if sent by the app.
+    TEMPLATE_VALIDITY_PERIOD = 12.hours
+    REQUIRED_ATTRIBUTES = %w[GivenName Surname EMailAddress].freeze # https://enmeshed.eu/integrate/attribute-values
+
+    def initialize(connector_url, api_key)
+      @api_key = api_key
+      @enmeshed_address = nil
+      @connector_url = connector_url
+      @conn = Faraday.new(connector_url, headers:)
+    end
+
+    def parse_result(response)
+      json = JSON.parse(response.body).deep_symbolize_keys
+
+      if json.include?(:error)
+        raise ConnectorError.new(
+          "Enmeshed connector response contained error: #{json[:error][:message]}. Full response was: #{response.body}"
+        )
+      end
+
+      json[:result]
+    rescue JSON::ParserError
+      raise ConnectorError.new("Enmeshed connector response could not be parsed. Received: #{response.body}")
+    end
+
+    def enmeshed_address
+      return @enmeshed_address if @enmeshed_address.present?
+
+      identity = parse_result @conn.get('/api/v2/Account/IdentityInfo')
+      @enmeshed_address = identity[:address]
+    end
+
+    def create_relationship_template(nbp_uid)
+      display_name_id = create_display_name
+
+      new_template = parse_result(@conn.post('/api/v2/RelationshipTemplates/Own') do |req|
+        req.body = relationship_template(display_name_id, nbp_uid).to_json
+      end)
+
+      Rails.logger.debug { "Enmeshed::ConnectorApi RelationshipTemplate created: #{new_template[:truncatedReference]}" }
+      new_template[:truncatedReference]
+    end
+
+    def create_display_name
+      new_attribute = parse_result(@conn.post('/api/v2/Attributes') do |req|
+        req.body = {content: display_name_attribute}.to_json
+      end)
+
+      new_attribute[:id]
+    end
```

Comment:
Yes, that works.

Reply:
Great to hear! Just one further question: When restarting the server, the memorization is lost, of course. This would also create a new attribute every time we restart it (at least with the current implementation), right?

Reply:
I added a check that queries the existing attributes and only creates a new display name attribute if there is none already.

Reply:
Thank you! I can indeed confirm that this is working as expected, and re-using the same attribute for multiple relationships. While testing, however, I found an issue with the current implementation:

It could happen that reusing doesn't work and instead returns an error with HTTP code `400`. Without making this up, I experienced the error myself by coincidence: With old data still present at my connector.

Consider the following example:

1. I start with a clean connector and no data being present.
2. First, I run the schoolmock-demo, creating the following attribute (before proceeding to create a `RelationshipTemplate`):
    ```json
    {
        "id": "ATTKxe0hOOyYk3N12KB5",
        "content": {
            "@type": "IdentityAttribute",
            "owner": "id15LtCHF3dvmseeNrBTMtk3dKmVPLEK5aEX",
            "value": {
                "@type": "DisplayName",
                "value": "Marie-Curie-Gymnasium"
            }
        },
        "createdAt": "2024-03-08T18:36:49.263Z"
    },
    ```
3. I scan the QR code and establish the relationship:
    ```json
    {
        "id": "ATTG0SerTvgDvrtl5qEo",
        "content": {
            "@type": "IdentityAttribute",
            "owner": "id15LtCHF3dvmseeNrBTMtk3dKmVPLEK5aEX",
            "value": {
                "@type": "DisplayName",
                "value": "Marie-Curie-Gymnasium"
            }
        },
        "createdAt": "2024-03-08T18:37:57.767Z",
        "shareInfo": {
            "peer": "id1E8j95kRUQ16WAtCtEFuBnQKb8hPRmcKgw",
            "requestReference": "REQW0CjYDBgE6FnsHNS4",
            "sourceAttribute": "ATTKxe0hOOyYk3N12KB5"
        }
    },
    ```
4. Some time later, I run the current branch for CodeHarbor. It does not create a new attribute, but selects the very first one that is present (in this case the one regarding the `Marie-Curie-Gymnasium`).
5. For the `RelationshipTemplate, the previously-selected attribute is re-used, but with another name. The relevant extract looks like this:
    ```json
	{
	    "@type": "ShareAttributeRequestItem",
	    "mustBeAccepted": true,
	    "attribute": {
	        "@type": "IdentityAttribute",
	        "owner": "id15LtCHF3dvmseeNrBTMtk3dKmVPLEK5aEX",
	        "value":{
	            "@type": "DisplayName",
	            "value": "CodeHarbor"
	        }
	    },
	    "sourceAttributeId": "ATTKxe0hOOyYk3N12KB5"
	}
	```
6. While noticing the re-use of `sourceAttributeId` and the new name for `DisplayName`, the connector responses with an error (HTTP code 400):
	```json
	{
	    "error": {
	        "code": "inheritedFromItem",
	        "message": "Some child items have errors.",
	        "id": "ERRU63UPGGETLWKDBWZ3",
	        "time": "2024-06-10T21:41:59.468Z",
	        "docs": "https://enmeshed.eu/integrate/error-codes#inheritedfromitem",
	        "stacktrace": [
	            "Error: Some child items have errors.",
	            "    at ValidationResult.fromItems (/usr/app/node_modules/@nmshd/consumption/dist/modules/common/ValidationResult.js:23:38)",
	            "    at OutgoingRequestsController.canCreate (/usr/app/node_modules/@nmshd/consumption/dist/modules/requests/outgoing/OutgoingRequestsController.js:32:60)",
	            "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
	            "    at async CreateOwnRelationshipTemplateUseCase.validateRelationshipTemplateContent (/usr/app/node_modules/@nmshd/runtime/dist/useCases/transport/relationshipTemplates/CreateOwnRelationshipTemplate.js:67:34)",
	            "    at async CreateOwnRelationshipTemplateUseCase.executeInternal (/usr/app/node_modules/@nmshd/runtime/dist/useCases/transport/relationshipTemplates/CreateOwnRelationshipTemplate.js:52:33)",
	            "    at async CreateOwnRelationshipTemplateUseCase.execute (/usr/app/node_modules/@nmshd/runtime/dist/useCases/common/UseCase.js:25:20)",
	            "    at async RelationshipTemplatesFacade.createOwnRelationshipTemplate (/usr/app/node_modules/@nmshd/runtime/dist/extensibility/facades/transport/RelationshipTemplatesFacade.js:29:16)",
	            "    at async RelationshipTemplatesController.createOwnTemplate (/usr/app/dist/modules/coreHttpApi/controllers/RelationshipTemplatesController.js:62:24)"
	        ]
	    }
	}
	```
7. I searched for the error code (and since it is not documented at all), only found the issue through some further trial-and-error.

My conclusion is that re-using an attribute only works as long as the string value is unchanged. Hence, we probably need to extend the filter for an existing attribute to check for the actual (and expected) value. Conclusively, we should create a new attribute if none of the existing ones matches our desired `DisplayName`.